### PR TITLE
Add support for custom templates in FSE themes

### PR DESCRIPTION
**Pull request by youknowriad**
This brings the "custom templates" feature from classic themes to FSE themes. While in classic themes, this  is done by adding a comment at the top of the template file, this can't be done for FSE themes since they are just HTML files.
(it might change later but let's do with what we have now)

So in order to support these, developers need to:
 -  Add custom files like `about.html`/ `custom.html` ... to the `block-templates` folder.
 - Add a `customTemplates` property to the root level of `theme.json` like so

```
	"customTemplates": {
		"custom": "My Custom Template"
	}
``` 

and that's it. The templates will show up in the template picker on the sidebar of the editor.
In terms of UI and integration with the template mode, we can do a lot more than the old select box. We can show a preview for templates and offer a modal or something like that for folks to pick from... We'd need some design thinking and explorations there. This PR just sets the technical requirements to make it work.

### CODE REVIEW

**Comment by gziolo** [comment]
This sounds great. It might require some alignment with how the default template types are structured:

https://github.com/WordPress/gutenberg/blob/5425353c950b57f16199cc6181f9103378216fc4/lib/full-site-editing/default-template-types.php#L16-L19

As of today it's:
- slug (as index)
- title 
- description

**Comment by jasmussen** [comment]
Wonderful. The exiting feature is called "Page Templates", referring to the fact that you can only use these when editing templates. It seems to make sense to rename from that since you could create a fresh new "Single" or "Archive" template with one of the custom ones as a starting point. But at the same time, "Custom Templates" may not be the best word, as they are still technically baked into the theme just like all the other templates stored in the `block-templates` directory — the real custom template, I would suggest, is created when a user modifies a template and saves that to the database.

I know the term is only used in the code, but is there a better term for it? "Additional templates"? "Unattached templates"? "Optional"?

**Comment by youknowriad** [comment]
I don't hold strong a opinion on the naming. The only places where this is used right now is `theme.json` file.  It can be just "templates" to be honest or maybe "extraTemplates" 🤷

**Comment by jasmussen** [comment]
I like "Templates".

**Comment by mtias** [comment]
> The exiting feature is called "Page Templates", referring to the fact that you can only use these when editing pages.

Custom templates are _by default_ assigned only to pages, but you can specify they should be made available to other post types: https://developer.wordpress.org/themes/template-files-section/page-template-files/#creating-page-templates-for-specific-post-types

I'm not sure we should use just "Templates" since all the other files (or posts in `wp_templates`) are also templates.

**Comment by jasmussen** [comment]
"Supplementary Templates" is accurate, but is a bit long and doesn't roll off the tongue. "Extra Templates" may be best then?

**Comment by mtias** [comment]
I'm confused, why not just page templates or custom templates?

**Comment by jasmussen** [comment]
I wouldn't object to either of those. But here's my reasoning.

- Page templates suggest they're just for pages, t but they could be starting points for anything.
- Custom templates are custom, sure, but not as custom as those a user saves to the database, and you can't delete them from the site editor like you can your other custom templates.

**Comment by mtias** [comment]
> but they could be starting points for anything

Not for anything, for different post types if so specified. I think it's important to have some continuity and not add to the burden of concepts.

**Comment by jasmussen** [comment]
Page Templates would probably be for for the SEO then.

**Comment by youknowriad** [comment]
Ok so I updated to use "postTemplates" as a config key and used the format suggested by @gziolo 
We can also set a "postTypes" property now to indicate on which CPTs the template applies.

File: lib/full-site-editing/page-templates.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Load the page templates in Gutenberg.
+ *
+ * @param array    $templates Page templates.
+ * @param WP_Theme $theme     WP_Theme instance.
+ * @param WP_Post  $post      The post being edited, provided for context, or null.
+ * @return array (Maybe) modified page templates array.
+ */
+function gutenberg_load_fse_page_templates( $templates, $theme, $post ) {
```
Comment (mtias) [inline lib/full-site-editing/page-templates.php:16]: not blocking, but we should start dropping the "fse" term in functions, etc, to something more general like `load_block_page_templates`

**Review by mtias — APPROVED** [review]

File: lib/full-site-editing/page-templates.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+			) {
+				$page_templates[ $key ] = $page_template['title'];
+			}
+		}
+	}
+
+	return $page_templates;
+}
+add_filter( 'theme_post_templates', 'gutenberg_load_block_page_templates', 10, 3 );
+add_filter( 'theme_page_templates', 'gutenberg_load_block_page_templates', 10, 3 );
```
Comment (swissspidy) [inline lib/full-site-editing/page-templates.php:42]: @youknowriad What about other post types? IMO the generic `theme_templates` filter should be used here.

**Comment by swissspidy** [comment]
> Ok so I updated to use "postTemplates" as a config key and used the format suggested by @gziolo
> We can also set a "postTypes" property now to indicate on which CPTs the template applies.

Why was `postTemplates` not kept? Why is it now `pageTemplates`?

WordPress uses the term "post template" internally (see `\WP_Theme::get_post_templates()` for example), so this feels like a step backward.

**Comment by mtias** [comment]
I think "post" can become quite confusing (particularly for a more user facing API, which building a theme is) because we already have templates that apply to the content of a post (on a CPT). Page signals more clearly it is applying beyond the content (the page template for a post type makes sense). Also the fact these templates are by default only available for post type = page. The fact that internally everything is a post doesn't necessarily mean, in my opinion, we should call something like menus postMenus on more surface level APIs. In any case, perhaps `customTemplates` is better here?

**Pull request merged**